src/librustdoc/html/render.rs

```diff
@@ -808,7 +808,16 @@ fn clean_srcpath<F>(src_root: &Path, p: &Path, keep_filename: bool, mut f: F) wh
     F: FnMut(&str),
 {
     // make it relative, if possible
-    let p = p.strip_prefix(src_root).unwrap_or(p);
+    let p = match p.strip_prefix(src_root) {
+        Ok(p) => p.to_path_buf(),
+        Err(_) => {
+            if p.is_absolute() {
+                Path::new("__root__").join(&p.strip_prefix(Path::new("/")).unwrap())
+            } else {
+                p.to_path_buf()
+            }
+        }
+    };
```

Comment:
This seems kinda sketchy, perhaps the iterator below could use `components` on `Path` to only look at the `Component::Normal` parts?